Src/libCZI/StreamsLib/curlhttpinputstream.cpp

```diff
@@ -0,0 +1,249 @@
+// SPDX-FileCopyrightText: 2023 Carl Zeiss Microscopy GmbH
+//
+// SPDX-License-Identifier: LGPL-3.0-or-later
+
+#include "curlhttpinputstream.h"
+#if LIBCZI_CURL_BASED_STREAM_AVAILABLE
+#include <cstdint>
+#include <string>
+#include <sstream>
+#include "../libCZI_StreamsLib.h"
+#include <curl/curl.h>
+
+using namespace std;
+using namespace libCZI;
+
+/*static*/void CurlHttpInputStream::OneTimeGlobalCurlInitialization()
+{
+    curl_global_init(CURL_GLOBAL_ALL);
+}
+
+/*static*/std::string CurlHttpInputStream::GetBuildInformation()
+{
+    const auto version_info = curl_version_info(CURLVERSION_NOW);
+
+    stringstream string_stream;
+    string_stream << "Version:" << version_info->version;
+    string_stream << " SSL:" << version_info->ssl_version;
+    if (version_info->age >= CURLVERSION_ELEVENTH)
+    {
+        for (size_t i = 0;; ++i)
+        {
+            if (version_info->feature_names[i] != nullptr)
+            {
+                if (i == 0)
+                {
+                    string_stream << " Features:";
+                }
+                else
+                {
+                    string_stream << ',';
+                }
+
+                string_stream << version_info->feature_names[i];
+            }
+            else
+            {
+                break;
+            }
+        }
+    }
+
+    return string_stream.str();
+}
+
+CurlHttpInputStream::CurlHttpInputStream(const std::string& url, const std::map<int, libCZI::StreamsFactory::Property>& property_bag)
+{
+    /* init the curl session */
+    CURL* curl_handle = curl_easy_init();
+    if (curl_handle == nullptr)
+    {
+        throw std::runtime_error("curl_easy_init() failed");
+    }
+
+    // put the handle into a smart-pointer with a custom deleter (so that we don't have to worry about calling curl_easy_cleanup in the lines below in case of error)
+    unique_ptr<CURL, void(*)(CURL*)> up_curl_handle(curl_handle, [](CURL* h)->void {curl_easy_cleanup(h); });
+
+    // construct a "curl-URL-handle" (an object that can be used to parse and manipulate URLs) -> https://curl.se/libcurl/c/libcurl-url.html
+    CURLU* curl_url_handle = curl_url();
+    if (curl_url_handle == nullptr)
+    {
+        throw std::runtime_error("curl_url() failed");
+    }
+
+    // put the url-handle also into a smart-pointer with a custom deleter
+    unique_ptr<CURLU, void(*)(CURLU*)> up_curl_url_handle(curl_url_handle, [](CURLU* h)->void {curl_url_cleanup(h); });
+
+    // Now, we parse the URL provided - the benefit of doing it this way is that at this point the URL is parsed for 
+    //  syntactical correctness, and we can be sure that the URL is valid. If we simply set the URL with CURLOPT_URL,
+    //  this would happen later on (when actually performing the request). I'd guess it is beneficial to catch this
+    //  kind of error early on.
+    // TODO(JBL): - we may want to restrict the schemes that are allowed 
```

Comment:
For non-ascii URLs we may just want to add sanitation which does as follows:
 - Check for non-ascii characters (see brief example: https://stackoverflow.com/a/48212993/22403590)
 - Convert non-ascii characters to hex (curl has curl_easy_escape for this : https://curl.se/libcurl/c/curl_easy_escape.html)

Reply:
Ja, I guess this is a matter of "better understanding the topic". I am not sure if we should just "require that the string is in a suitable" format or whether we should do some massaging here. I'd opt for - let's wait and see here, at least I am not sure what the "best solution" here would have to look like.

Reply:
I'd recommend the simple solution to always escape as a safety pre-caution

Reply:
I added some notes on this topic in the code. I'd vote for "leave as is for the time being" here :-)